### PR TITLE
insert: fix the auto id retry won't cast the datum to origin type (#29997)

### LINKED ISSUE #29892: Retrying transaction with a double type auto-increment column leads to data error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Prepare:
```sql
tidb> set global tidb_txn_mode='optimistic';
Query OK, 0 rows affected (0.01 sec)

tidb> set global tidb_disable_txn_auto_retry=false;
Query OK, 0 rows affected (0.01 sec)
```

In session1:
```sql
session1> create table t(a double key auto_increment, b int);
Query OK, 0 rows affected (0.03 sec)
session1> insert into t values (146576794, 1);
Query OK, 1 row affected (0.01 sec)
session1> begin;
session1> insert into t(b) select 1;
```

In session2:
```sql
session2> begin;
Query OK, 0 rows affected (0.00 sec)
session2> insert into t values (146576795, 1);
Query OK, 1 row affected (0.00 sec)
session2> insert into t values (146576796, 1);                             
Query OK, 1 row affected (0.00 sec)
session2> commit;
Query OK, 0 rows affected (0.01 sec)
```

In session1:
```sql
session1> commit;
Query OK, 0 rows affected (2.63 sec)
session1> select * from t;
ERROR 1105 (HY000): Failed to decode datum payload as f64
```

### 2. What did you expect to see? (Required)

Query without an error.

### 3. What did you see instead (Required)

```
ERROR 1105 (HY000): Failed to decode datum payload as f64
```

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #29997 to release-5.3
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/30023
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/30023:release-5.3-51f53f3376bd
```

---

Signed-off-by: ailinkid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close #29892

### What is changed and how it works?

- Fix the session retry won't cast the datum to the origin type under `insert into select from`
- In the original issue #29892, the session1 commit should fail with `duplicate entry` while it doesn't. Because the session2's inserted key is built from (double)146576795, while the session1's retried key is built from (int64)146576795 without cast. Different type contributing the comparison failure.
 

Tests 

- [x] Unit test
- [x] Manual test (add detailed scripts or steps below)
```
same as what has been described in the issue
```

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by guo-shaoge** [comment]
@AilinKid please continue to work on this, v5.3 is about to realease a new version.

File: executor/insert_test.go [hunk 1]
```diff
@@ -1783,3 +1783,73 @@ func (s *testSuite13) TestIssue26762(c *C) {
 	_, err = tk.Exec("insert into t1 values('2020-02-31');")
 	c.Assert(err.Error(), Equals, `[table:1292]Incorrect date value: '2020-02-31' for column 'c1' at row 1`)
 }
+<<<<<<< HEAD
+=======
```
Comment (djshow832) [inline executor/insert_test.go:1787]: Please fix conflicts.

File: executor/insert_test.go [hunk 2]
```diff
@@ -1783,3 +1783,73 @@ func (s *testSuite13) TestIssue26762(c *C) {
 	_, err = tk.Exec("insert into t1 values('2020-02-31');")
 	c.Assert(err.Error(), Equals, `[table:1292]Incorrect date value: '2020-02-31' for column 'c1' at row 1`)
 }
+<<<<<<< HEAD
```
Comment (zimulala) [inline executor/insert_test.go:1786]: Please fix these conflicts.

**Review by djshow832 — APPROVED** [review]

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**